[Site Isolation] Remote iframe and main frame media state can overwrite each other

bnham opened the pull request:
#### 4b848363d877ca2db335de6505dacb37e120d66e
<pre>
[Site Isolation] Remote iframe and main frame media state can overwrite each other
<a href="https://bugs.webkit.org/show_bug.cgi?id=287641">https://bugs.webkit.org/show_bug.cgi?id=287641</a>
<a href="https://rdar.apple.com/144739815">rdar://144739815</a>

Reviewed by Charlie Wolfe.

Right now, the only place where we store media state in the UIProcess for a given web page is in
WebPageProxy::Internals::mediaState. Both the main frame and remote frames processes overwrite that
same media state in in response to the WebPageProxy::isPlayingMediaDidChange IPC. So a remote frame
stopping media playback could erroneously clear the playback flag in media state even if the main
frame is still playing back media.

To fix this, make WebPageProxy track the media state for the main frame and RemotePageProxy track
the media state for the OOP frames it owns.

I also had to make PlaybackTargetClientContextIdentifier process qualified for the tests to work,
since the main frame and remote frame were generating clashing identifiers and tripping an assert in
`WebMediaSessionManager::addPlaybackTargetPickerClient`.

* Source/WebCore/Modules/airplay/PlaybackTargetClientContextIdentifier.h:
* Source/WebCore/Modules/airplay/WebMediaSessionManager.cpp:
(WebCore::WebMediaSessionLogger::logAlways const):
(WebCore::WebMediaSessionManager::addPlaybackTargetPickerClient):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/ProcessQualified.serialization.in:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::~RemotePageProxy):
(WebKit::RemotePageProxy::didReceiveMessage):
(WebKit::RemotePageProxy::isPlayingMediaDidChange):
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::mediaState const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::Internals::Internals):
(WebKit::WebPageProxy::~WebPageProxy):
(WebKit::WebPageProxy::close):
(WebKit::WebPageProxy::isNoLongerAssociatedWithRemotePage):
(WebKit::WebPageProxy::resetState):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
(WebKit::WebPageProxy::isPlayingMediaDidChange):
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
(WebKit::WebPageProxy::updatePlayingMediaDidChangeTimerFired):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::updateAudibleMediaAssertions):
(WebKit::WebProcessProxy::updateMediaStreamingActivity):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::isPlayingMediaDidChange):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::callMethodOnFirstVideoElementInFrame):
(TestWebKitAPI::TEST(SiteIsolation, PlayAudioInMultipleFrames)):
(TestWebKitAPI::TEST(SiteIsolation, PlayAudioInRemoteFrameThenRemove)):

Canonical link: <a href="https://commits.webkit.org/290482@main">https://commits.webkit.org/290482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4344fe123d77dba180f97f989ac7d72f7b929d8a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90166 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9695 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45082 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95167 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40942 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92218 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10082 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18004 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69416 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27016 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93167 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7723 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81793 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49777 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/89667 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7447 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36164 "Found 60 new test failures: compositing/animation/repaint-after-clearing-shared-backing.html fast/workers/dedicated-worker-lifecycle.html fonts/monospace.html fonts/sans-serif.html fonts/serif.html fonts/unicode-character-font-crash.html fonts/use-typo-metrics-1.html http/tests/eventsource/eventsource-page-cache-connected.html http/tests/eventsource/eventsource-page-cache-connecting.html http/tests/security/contentSecurityPolicy/block-all-mixed-content/insecure-script-in-iframe.html ... (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40073 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77779 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37220 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96992 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17354 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12754 "Found 2 new test failures: http/wpt/mediarecorder/MediaRecorder-matroska-AV-audio-video-dataavailable.html imported/w3c/web-platform-tests/service-workers/service-worker/xhr-content-length.https.window.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78413 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17611 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77617 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77618 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19160 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22071 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20670 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10606 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17364 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22690 "") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17105 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20557 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18889 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/RemotePageProxy.cpp:
```diff
@@ -256,4 +264,23 @@ WebProcessActivityState& RemotePageProxy::processActivityState()
+void RemotePageProxy::isPlayingMediaDidChange(WebCore::MediaProducerMediaStateFlags newState, WebCore::FrameIdentifier mainFrameID)
+{
+    m_mediaState = newState;
+    m_mediaStateMainFrameID = mainFrameID;
```
The reason I added this main frame ID is because I don't really know how the lifetime of RemotePageProxy within the BrowsingContextGroup relates to its WebPageProxy when the WebPageProxy's state is being reset (e.g. in `WebProcessProxy::resetState` which can occur after the main frame's WebContent process terminates).

Basically, I didn't want the main WebPageProxy to take this RemotePageProxy's mediaState in account if its mainFrame is null (e.g. just got closed) or doesn't match the current main frame ID in the WebPageProxy (e.g. due to PSON).

I'm not sure this is the right/best way to do this.

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -1796,7 +1799,6 @@ void WebPageProxy::close()
     internals().audibleActivityTimer.stop();
 
     stopAllURLSchemeTasks();
-    updatePlayingMediaDidChange(MediaProducer::IsNotPlaying);
```
I removed this because it seemed like it was doing nothing--we are already calling it in `resetState` above.

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11278,8 +11286,6 @@ void WebPageProxy::resetStateAfterProcessExited(ProcessTerminationReason termina
         protectedPageLoadState->reset(transaction);
     }
 
-    updatePlayingMediaDidChange(MediaProducer::IsNotPlaying);
```
I removed this because it seemed like it was doing nothing--we are already calling it in `resetState` above.

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13372,8 +13382,16 @@ MediaProducerMediaStateFlags WebPageProxy::reportedMediaState() const
+    if (m_mainFrame) {
+        protectedBrowsingContextGroup()->forEachRemotePage(*this, [&newState, mainFrameID = m_mainFrame->frameID()](auto& remotePageProxy) {
+            if (auto remotePageMediaState = remotePageProxy.mediaStateIfMatchesMainFrameID(mainFrameID))
+                newState.add(*remotePageMediaState);
```
I need to talk to media people to see if or'ing the flags like this makes sense. It seems like it should, since we already do this on the WebProcess side:

```
void Page::updateIsPlayingMedia()
{
    MediaProducerMediaStateFlags state;
    forEachDocument([&](auto& document) {
        state.add(document.mediaState());
    });
```

eric-carlson replied:
I agree, I think this is the right approach.

bnham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -7445,6 +7447,11 @@ void WebPageProxy::processIsNoLongerAssociatedWithPage(WebProcessProxy& process)
 
+void WebPageProxy::isNoLongerAssociatedWithRemotePage(RemotePageProxy&)
+{
+    internals().updatePlayingMediaDidChangeTimer.startOneShot(0_s);
```
I added this timer to coalesce calls to `updatePlayingMediaDidChange` if a bunch of remote pages are destroyed during the same event loop turn.

Possibly an unnecessary optimization.

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.messages.in:
```diff
@@ -467,7 +467,7 @@ messages -> WebPageProxy {
     IsAnyAnimationAllowedToPlayDidChange(bool anyAnimationCanPlay)
 #endif
-    IsPlayingMediaDidChange(OptionSet<WebCore::MediaProducerMediaState> state)
+    IsPlayingMediaDidChange(OptionSet<WebCore::MediaProducerMediaState> state, WebCore::FrameIdentifier mainFrameID)
```
`mainFrameID` is unused now, should be removed.

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13372,8 +13382,15 @@ MediaProducerMediaStateFlags WebPageProxy::reportedMediaState() const
+        protectedBrowsingContextGroup()->forEachRemotePage(*this, [&newState](auto& remotePage) {
+            newState.add(remotePage.mediaState());
+        });
+    }
```
```suggestion
    protectedBrowsingContextGroup()->forEachRemotePage(*this, [&newState](auto& remotePage) {
        newState.add(remotePage.mediaState());
    });
```

charliewolfe commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -3881,4 +3881,95 @@ HTTPServer server({
+    bool isNotPlayingAudio = TestWebKitAPI::Util::waitFor([view = webView]() {
+        return ![view _isPlayingAudio];
+    });
+    EXPECT_TRUE(isNotPlayingAudio) << "Should not be playing audio after removing iframe";
```
I suspect these tests might be flaky right now. Instead of waiting for the JavaScript evaluation to finish, you could just wait for the correct state. This might also fix the tests on iOS.
```suggestion
    [webView evaluateJavaScript:@"document.getElementsByTagName('video')[0].play()" inFrame:[webView firstChildFrame] completionHandler:nil];
    while (![webView _isPlayingAudio])
        Util::spinRunLoop();

    [webView evaluateJavaScript:@"document.querySelectorAll('iframe').forEach(iframe => iframe.remove())" completionHandler:nil];
    while ([webView _isPlayingAudio])
        Util::spinRunLoop();
```

bnham committed:
[Site Isolation] Remote iframe and main frame media state can overwrite each other
https://bugs.webkit.org/show_bug.cgi?id=287641
rdar://144739815

Reviewed by Charlie Wolfe.

Right now, the only place where we store media state in the UIProcess for a given web page is in
WebPageProxy::Internals::mediaState. Both the main frame and remote frames processes overwrite that
same media state in in response to the WebPageProxy::isPlayingMediaDidChange IPC. So a remote frame
stopping media playback could erroneously clear the playback flag in media state even if the main
frame is still playing back media.

To fix this, make WebPageProxy track the media state for the main frame and RemotePageProxy track
the media state for the OOP frames it owns.

I also had to make PlaybackTargetClientContextIdentifier process qualified for the tests to work,
since the main frame and remote frame were generating clashing identifiers and tripping an assert in
`WebMediaSessionManager::addPlaybackTargetPickerClient`.

* Source/WebCore/Modules/airplay/PlaybackTargetClientContextIdentifier.h:
* Source/WebCore/Modules/airplay/WebMediaSessionManager.cpp:
(WebCore::WebMediaSessionLogger::logAlways const):
(WebCore::WebMediaSessionManager::addPlaybackTargetPickerClient):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/ProcessQualified.serialization.in:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::~RemotePageProxy):
(WebKit::RemotePageProxy::didReceiveMessage):
(WebKit::RemotePageProxy::isPlayingMediaDidChange):
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::mediaState const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::Internals::Internals):
(WebKit::WebPageProxy::~WebPageProxy):
(WebKit::WebPageProxy::close):
(WebKit::WebPageProxy::isNoLongerAssociatedWithRemotePage):
(WebKit::WebPageProxy::resetState):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
(WebKit::WebPageProxy::isPlayingMediaDidChange):
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
(WebKit::WebPageProxy::updatePlayingMediaDidChangeTimerFired):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::updateAudibleMediaAssertions):
(WebKit::WebProcessProxy::updateMediaStreamingActivity):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::isPlayingMediaDidChange):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::callMethodOnFirstVideoElementInFrame):
(TestWebKitAPI::TEST(SiteIsolation, PlayAudioInMultipleFrames)):
(TestWebKitAPI::TEST(SiteIsolation, PlayAudioInRemoteFrameThenRemove)):

Canonical link: https://commits.webkit.org/290482@main
